packages/@vue/cli-shared-utils/lib/env.js

```diff
@@ -79,32 +79,33 @@ exports.hasProjectGit = (cwd) => {
 }
 
 let _hasPnpm
-let _hasPnpm3orLater
+let _pnpmVersion
 const _pnpmProjects = new LRU({
   max: 10,
   maxAge: 1000
 })
 
-exports.hasPnpm3OrLater = () => {
-  if (process.env.VUE_CLI_TEST) {
-    return true
-  }
-  if (_hasPnpm3orLater != null) {
-    return _hasPnpm3orLater
-  }
+function getPnpmVersion () {
   try {
-    const pnpmVersion = execSync('pnpm --version', {
+    _pnpmVersion = execSync('pnpm --version', {
       stdio: ['pipe', 'pipe', 'ignore']
     }).toString()
     // there's a critical bug in pnpm 2
     // https://github.com/pnpm/pnpm/issues/1678#issuecomment-469981972
     // so we only support pnpm >= 3.0.0
     _hasPnpm = true
-    _hasPnpm3orLater = semver.gte(pnpmVersion, '3.0.0')
-    return _hasPnpm3orLater
-  } catch (e) {
-    return (_hasPnpm3orLater = false)
+  } catch (e) {}
+  return _pnpmVersion || '0.0.0'
+}
+
+exports.hasPnpmXOrLater = (version) => {
+  if (process.env.VUE_CLI_TEST) {
+    return true
+  }
+  if (_pnpmVersion != null) {
+    return semver.gte(_pnpmVersion, version)
```

Comment:
This logic should be moved to `getPnpmVersion` because of separation of concerns. The `_pnpmVersion` cache is an implementation detail of `getPnpmVersion()`.

And separation of concerns *is* required here, because, if not for such consideration, the two functions can essentially be merged into one.